File: Day 065/Problem 140. Word Break II.cpp
class Solution {
public:
    
    vector<string> result;
    void solve(int index, const string &s, string &combination, const unordered_set<string> &dict) {
        if (index == s.size()) {
            result.push_back(combination);
            return;
        }

        for (int i = index; i < s.size(); i++) {
            string str = s.substr(index, i - index + 1);
            if (dict.find(str) != dict.end()) {
                if (combination.size() == 0) combination += str;
                else {
                    combination += " " + str;
                }

                solve(i + 1, s, combination, dict);

                for (int c : str) {
                    combination.pop_back();
                }

                if (combination.size() != 0) combination.pop_back();
            }
        }
    }

    vector<string> wordBreak(string s, vector<string>& wordDict) {
        string combination = "";
        unordered_set<string> dict(wordDict.begin(), wordDict.end());
        solve(0, s, combination, dict);

        return result;
    }
};

File: Day 091/Problem 140.Word Break II.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

class Solution {
    List<String> sentences = new ArrayList<>();

    private void dfs(int index, String s, StringBuilder sentence, Set<String> wordSet) {
        if (index == s.length()) {
            sentences.add(sentence.toString().trim());
            return;
        }

        for (int i = index + 1; i <= s.length(); i++) {
            String substr = s.substring(index, i);
            if (wordSet.contains(substr)) {
                int len = sentence.length();

                sentence.append(substr).append(" ");
                dfs(i, s, sentence, wordSet);
                sentence.setLength(len);
            }
        }
    }

    public List<String> wordBreak(String s, List<String> wordDict) {
        Set<String> wordSet = new HashSet<>(wordDict);
        dfs(0, s, new StringBuilder(), wordSet);
        return sentences;
    }
}